### PR TITLE
Landing detector: replace mc_min_fly_throttle with average throttle

### PR DESCRIPTION
**Pull request by theArchLadder**
Improving one aspect of issue: #219 

Currently we are using  `mc_min_fly_throttle` to set a throttle value to compare current throttle to. This has a few problems IMHO:
1. Stupid users (like me) may set it wrong (look at me, i had it too high and got a disarm-in-air!).
2. In one of @stronnag logs it looked like the hover throttle went from ~1550 to ~1650 with full to empty battery, this gives inconsistent behavior no matter how well tuned mc_min_fly_throttle is.
3. What if one uses a heaver or light battery? Add or remove a gimbal or change the weigth in any other way?
4. Impossible to set good default. Either to high so racing quads risk disarming in air, or to low so it takes forever to disarm big quads.
5. It requires tuning.

I got an idea that may solve all of those problems. I have started playing around with measuring the average throttle during descent and comparing the current throttle to that. I have flown with just a 40ms timeout, and my quad disarms instantly on touchdown, before the landing legs has even settled in to the grass, but the throttle check has never passed in-air! Not even close actually.

This needs a little bit more testing and cleanup before it's ready to merge, but i'm really happy with the results so far, better than i expected :smiley: Comments and feedback are welcome as usual!

### CODE REVIEW

**Comment by theArchLadder** [comment]
Lets look at some logs.
'`navDebug[1] = (*landingThrSum / *landingThrSamples) - rcCommandAdjustedThrottle;`
So a positive value means that the average throttle is above current throttle, i should probably have inverted this but i'm to lazy to redo it.
First image is before the fake-starting-average, second one is with that feature.

The throttle is at hover when the descent starts, and then quickly drops a bit to get to -200cm/s, there is a small risk of a false-positive detection here, the highest i have seen is 38 and it needs to go over 50 to pass, but still, i wanted to make this safer so i start with a fake low average so now it's below 0 all the time and then shoots up to 120 on touchdown. The line in the graph looks very clean and safe now imho. And it still disarms even if RTH is activated on the ground.

Here is what it looks like in blackbox log.
navDebug[1] is the interesting one. Second graph used a longer timeout (navDebug[2])

scrot_20160527-113944

scrot_20160527-112951

**Comment by stronnag** [comment]
There's a couple of logs at http://seyrsnys.myzen.co.uk/inav_pr249_2016-05-27/ with the dif on top of near-rth-slow-descent-fix branch. In each case, disarm was in 2 or 3 seconds on landing. Very impressive. Please can we have this merged (into near-rth-slow-descent-fix branch if possible, as it's logically related and make building easy).

It gets flown on the precious tricopter tomorrow; I can offer no higher praise.

**Comment by theArchLadder** [comment]
@stronnag A huge thanks for helping out by providing those logs!
I just did a quick review of them and it looks good, every landing looks consistent and it's similar to my quad.
It is a bit "to safe" though so the timer gets reset on the bounce after touchdown (it only adds a few 100ms), and the "fake starting average" has a bit more weight than needed to it, so i think i will fine tune some stuff here, but still, i'm really happy with how this turned out!

TODO before redo to merge:
1. i will fine tune based on stronnag and my logs.
2. i use int_t, it would make more sense to use uint_t but i get a warning on build, have to figure out proper solution. Sample-variable have a rollover at 10 minutes i think, should we waste another byte to be safe?
3. it will suck when someone say "auto-disarm doesn't work" and we have no way of debugging. Would be nice to have a blackbox variable for this. I just have to figure out how that works...

**Comment by digitalentity** [comment]
I'll add blackbox logging when you say the code is ready.

File: src/main/flight/navigation_rewrite_multicopter.c [hunk 1]
```diff
@@ -481,13 +481,24 @@ bool isMulticopterLandingDetected(uint32_t * landingTimer, bool * hasHadSomeVelo
...
     // We use rcCommandAdjustedThrottle to keep track of NAV corrected throttle (isLandingDetected is executed
     // from processRx() and rcCommand at that moment holds rc input, not adjusted values from NAV core)
     bool minimalThrust = rcCommandAdjustedThrottle < posControl.navConfig->mc_min_fly_throttle;
-    
+    */
+
+    // We have likely landed if throttle is 50 units below average descend throttle
+    *landingThrSamples += 1;
+    *landingThrSum += rcCommandAdjustedThrottle;
+    bool minimalThrust= rcCommandAdjustedThrottle < *landingThrSum / *landingThrSamples - 50;
```
Comment (theArchLadder) [inline src/main/flight/navigation_rewrite_multicopter.c]: Help wanted!
If the landingThr-variables are unsigned integers, i get this warning:
```
./src/main/flight/navigation_rewrite_multicopter.c:489:52: 
warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     bool minimalThrust= rcCommandAdjustedThrottle < *landingThrSum / *landingThrSamples - 40;
                                                    ^
```
I could solve that by using this code:
`bool minimalThrust= rcCommandAdjustedThrottle < (int16_t)(*landingThrSum / *landingThrSamples - 40);`
or just use signed intergers for landingThr-variables. @digitalentity how would you like it?

Also, a int16_t/uint16_t is only good for 10/20 minutes of sample counting, should i upgrade to 32 bit integer?

File: src/main/flight/navigation_rewrite_multicopter.c [hunk 2]
```diff
@@ -481,13 +481,22 @@ bool isMulticopterLandingDetected(uint32_t * landingTimer, bool * hasHadSomeVelo
...
 
-    // Throttle should be low enough
+    // We have likely landed if throttle is 40 units below average descend throttle
     // We use rcCommandAdjustedThrottle to keep track of NAV corrected throttle (isLandingDetected is executed
     // from processRx() and rcCommand at that moment holds rc input, not adjusted values from NAV core)
-    bool minimalThrust = rcCommandAdjustedThrottle < posControl.navConfig->mc_min_fly_throttle;
-    
+    *landingThrSamples += 1;
+    *landingThrSum += rcCommandAdjustedThrottle;
+    bool minimalThrust= rcCommandAdjustedThrottle < *landingThrSum / *landingThrSamples - 40;
```
Comment (theArchLadder) [inline src/main/flight/navigation_rewrite_multicopter.c]: Help wanted!
If the landingThr-variables are unsigned integers, i get this warning:

```
./src/main/flight/navigation_rewrite_multicopter.c:489:52: 
warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     bool minimalThrust= rcCommandAdjustedThrottle < *landingThrSum / *landingThrSamples - 40;
                                                    ^
```

I could solve that by using this code:
`bool minimalThrust= rcCommandAdjustedThrottle < (int16_t)(*landingThrSum / *landingThrSamples - 40);`
or just use signed intergers for landingThr-variables. @digitalentity how would you like it?

File: src/main/flight/navigation_rewrite.c [hunk 3]
```diff
@@ -1626,7 +1633,7 @@ bool isLandingDetected(void)
         landingDetected = isFixedWingLandingDetected(&landingTimer);
     }
     else {
-        landingDetected = isMulticopterLandingDetected(&landingTimer, &hasHadSomeVelocity);
+        landingDetected = isMulticopterLandingDetected(&landingTimer, &hasHadSomeVelocity, &landingThrSum, &landingThrSamples);
```
Comment (digitalentity) [inline src/main/flight/navigation_rewrite.c:1636]: I suggest moving all these parameters to a `typedef struct` to make things cleaner.

File: src/main/flight/navigation_rewrite_multicopter.c (hunk 2 above)
Comment (digitalentity) [inline src/main/flight/navigation_rewrite_multicopter.c]: For better readability I suggest renaming `minimalThrust` to `isAtMinimalThrust`. This is a true-falce flag, not the actual value of minimal thrust.

**Comment by digitalentity** [comment]
@theArchLadder we need to better test this approach on various sized machines, but I like the general idea.

File: src/main/flight/navigation_rewrite.c [hunk 4]
```diff
@@ -1611,11 +1611,18 @@ void calculateFarAwayTarget(t_fp_vector * farAwayPos, int32_t yaw, int32_t dista
...
 
 void resetLandingDetector(void)
 {
     landingTimer = micros();
     hasHadSomeVelocity = false;
+    
+    // When descent starts the throttle is at hover and quickly drops to gain descend velocity,
+    // Start with a fake low throttle average to avoid passing the test on that throttle drop.
+    landingThrSum = 8 * 1000;
+    landingThrSamples = 8;
```
Comment (digitalentity) [inline src/main/flight/navigation_rewrite.c:1625]: This should be calculated using time and update rate. I don't like magic numbers :smile: \
However for now it is good enough!

File: src/main/flight/navigation_rewrite_multicopter.c (hunk 2 above)
Comment (digitalentity) [inline src/main/flight/navigation_rewrite_multicopter.c]: I don't like the `40` magic constant. We should probably make it a percentage of `(*landingThrSum / *landingThrSamples - minthrottle)`. What do you think?

**Comment by theArchLadder** [comment]
@digitalentity Thanks! Looks good on my 250 with default and very non-default althold settings, it also looked good (and very similar to mine) on stronnag's bigger quad, but i agree, more testing would be good :+1:

File: src/main/flight/navigation_rewrite.c (hunk 4 above)
Comment (theArchLadder) [inline src/main/flight/navigation_rewrite.c:1625]: I agree it's a bit of a ugly hack (that happens to work beautiful). Making it independent of the 50hz update rate for landing timer would be good though.

File: src/main/flight/navigation_rewrite_multicopter.c (hunk 2 above)
Comment (theArchLadder) [inline src/main/flight/navigation_rewrite_multicopter.c]: Renamed minimalThrust to isAtMinimalThrust.

Hmm, let me see if i understand your percentage thingy right:
My minthrottle=1040, hover=1220, so you formula would give 1220-1040=180, lets say current throttle is at 1180-1040=140, 140/180=77%, was it something like this you meant? That would require a bigger change in absolute throttle value with higher hover though, not sure if we want that...?

Comment (digitalentity) [inline src/main/flight/navigation_rewrite_multicopter.c]: My idea is that bigger hover throttle usually means bigger throttle variation to maintain velocity. On a quad with 1700 hover that hardcoded 40 threshold might be exceeded during normal descent (i.e. caused by a gust of wind).

Comment (theArchLadder) [inline src/main/flight/navigation_rewrite_multicopter.c]: I compared my and stronnags bblogs, I looked at the throttle dip on descent start.
My copter reach 0.73g with throttle at 67% of hover(1220), or 60 units under.
Stronnag's copter reached 0.73g with throttle at 79% of hover(), or 100 units under.

But the thing is, on touchdown the throttle change looks almost identical in absolute numbers between my and stronnag's copter. I guess the althold controller works with throttle units, not percentage or anything like that...?

Comment (theArchLadder) [inline src/main/flight/navigation_rewrite_multicopter.c]: Hmm, i agree that magic numbers suck... Could we calculate what the vel_p output would be with 50cm/s vertical velocity error? Looks like vel_p does most of the work to bring down throttle right after touchdown.

**Comment by stronnag** [comment]
There's another log file http://seyrsnys.myzen.co.uk//inav_pr249_2016-05-28/LOG01152.TXT. This time from the 600mm TRI. Pretty much same effect; disarms reliably with 2-3 seconds.

**Comment by theArchLadder** [comment]
@stronnag Thanks! Looks like you didn't use the commit 
"Landing detector fine-tuning and cleanup           6da0d51"
Logs tell me it would have had a little more consistent and accurate disarm with that, but the logs was still useful!

flight 2 and 3 was from more than 16 meters of altitude, but flight one was only from 3 meters of altitude, and on that flight the "fake starting average" was not really out-weight by the true average, so it took a little longer to disarm. Bad.

I'm now starting to think that perhaps it would be better to wait until hasHadSomeVelocity is true, and then start to measure average throttle. At that point the throttle is almost at its minimum during the descent...
If that works we could get rid of that ugly fake starting average... It's raining here now but i will try to test it tomorrow.

**Comment by stronnag** [comment]
It was the same build that I flew yesterday (for consistency, or I'm too damn lazy to patch -R and re-patch). The first attempt was merely to limit the damage in the unlikely event that it didn't work on this frame :)

**Comment by digitalentity** [comment]
@theArchLadder I guess it's time to make landing detector a proper state machine. I'll create a branch and merge your work into it so more devs could improve it (myself included).
We also need to cover the case when hasHadSomeVelocity condition is never satisfied (too low and landed before it gained enough vel.

**Comment by theArchLadder** [comment]
@digitalentity Sounds good. I know my programming skills are a bit crappy and it ends up with a bunch of annoying noob-programming-questions. Perhaps it would be better if you would just apply your magic touch on the code directly instead of having to explain coding-stuff to me...? I can still learn from the commits.
Still,  i hope you guys like my work with coming up with ideas, making ugly code to test it, tweak it and test it again...

**Comment by theArchLadder** [comment]
Had some fun with instant disarm today
https://www.youtube.com/watch?v=KPuE-l3v4xQ
It is very consistent, every landing looks like this, and no accidents. I'm impressed :smile:

**Comment by theArchLadder** [comment]
@digitalentity Do you have something in mind for when hasHadSomeVelocity never passes?

**Comment by theArchLadder** [comment]
I suppose waiting until "target velocity is not increasing" would have almost the same effect but it will pass even if RTH is activated on the ground... But this would mean that the landing detector is dependant on the current behavior of target velocity, not sure if that's much better...

**Comment by digitalentity** [comment]
Closing this one. Work should continue in #256.
@theArchLadder you now have commit rights to INAV repo. Feel free to create branches and work in the INAV repository :smile:

**Pull request closed without merging**